File: Data Structures & Algorithms/count-number-of-islands/submission-0.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
      int count = 0;
      vector<vector<bool>> visited(grid.size(), vector<bool>(grid[0].size(), false));
        for(int i = 0;i< grid.size(); i++){
          for(int j = 0;j< grid[0].size();j++){
            if(grid[i][j] == '1'&&visited[i][j]==false){
              dfs(i,j,grid,visited);
              count++;
            }
          }
        }
        return count;
    }

  private:
     void dfs(int i,int j, vector<vector<char>>& grid, vector<vector<bool>>& visited){
      if(i<0 || j<0 || i>=grid.size() || j>= grid[0].size() || visited[i][j] == true
      || grid[i][j] == '0'){
        return;
      }
       
       visited[i][j] = true;
       dfs(i,j-1,grid,visited);
       dfs(i,j+1,grid,visited);
       dfs(i+1,j,grid,visited);
       dfs(i-1,j,grid,visited);

     }
};

File: Data Structures & Algorithms/count-connected-components/submission-0.java
class Solution {
    public int countComponents(int n, int[][] edges) {
        ArrayList<ArrayList<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }

        for (int edge[] : edges) {
            int u = edge[0];
            int v = edge[1];
            adj.get(u).add(v);
            adj.get(v).add(u);
        }
        int count = 0;
       int visited[] = new int[n];
        for(int i = 0;i<n;i++){
            if(visited[i]==0){
                dfs(adj,i,visited);
                count++;
            }
        }
        return count;

    }

    public void dfs(ArrayList<ArrayList<Integer>> adj, int u, int visited[]) {
        visited[u] = 1;
        for (int v : adj.get(u)) {
            if (visited[v] == 0) {
                dfs(adj, v, visited);
            }
        }
    }
}
